doc/Lorre's Block Fetching Algorithm

```diff
@@ -0,0 +1,96 @@
+we end our fork checking here. This is the base or neutral case.
+
+Case 2: Block at level m-1 is not in the blocks table, and is not in the invalidated blocks table.
+
+We've found a block that is not accounted for. If there are any blocks in the blocks table that have
+the same level as this block, we should make sure those blocks are written to the invalidated blocks
+table with the is_invalidated flag set to true. Then, this block will be collected to be written
+to the Blocks table.
+
+Case 3: Block at level m-1 is in the blocks table, and is also in the invalidated blocks table.
+
+We've found a block we've seen before that has already been invalidated at least once. We'll
+simply make sure that this particular block in the invalidated blocks table is has
+its is_invalidated flag set to false, and every other block at the same level will be set
+to true. Since it's already in the blocks table, no block writing is necessary.
+
+INVARIANT: at any given point in time, there will only be at most one block at a given level
+with is_invalidated set to false in the invalidated blocks table. [iii]
+
+Case 4: Block at level m-1 is not in the blocks table, but is in the invalidated blocks table.
+
+IMPOSSIBLE. This function is the only function that writes to the invalidated blocks table, and will only
+write a block to the table if the block is already in the table and passes the criteria to be invalidated.
+
+We then continue the same checks for blocks m-2, m-3...until Case 1, our base case, has been reached. Then,
+we are done collecting forked blocks for writing, as well as keeping our invalidated blocks table up to speed.
+
+---------------------------------------------------------------------------------------------------------
+
+As long as assumptions are valid and the invariant is maintained, this algorithm will work as intended:
+
+[i] In a non-forked chain, there can only be one block at each level.
+
+[ii] Internal consistency with our database is all we can hope for, as there is no notion of a
+     globally correct chain.
+
+[iii] At any given point in time, there will only be at most one block at a given level
+      with is_invalidated set to false in the invalidated blocks table
+
+-----------------------------------------------------------------------------------------------------------
+
+Spec-ing Unit Tests for the Forking Algorithm:
+
+GetBlocks should correctly process the following chains:
+
+i) A regular, unforked chain
+
+ii) A chain with one fork, or two branches
+
+iii) A chain with one fork, and Lorre's iterations happen to go back and forth on either one of these chains.
+
+iv) A chain with two forks, or three branches, ending on the second fork/third branch
+
+v) A chain with two forks, or three branches, and Lorre's iterations go back and forth between all three of
+these chains, until it's decided that the second branch is the true internal chain for an extended period of time.
```

Comment:
how do we decide about a long-term internal chain? what would be the conditions?

Reply:
See above definition of an internally consistent database mapping of the blockchain